### ISSUE TITLE
Loss and Error surface

### ISSUE THREAD

**Issue by pillarliang** [pillarliang/blog#8]
# Introdution
*损失函数*（loss function）能够量化目标的*实际*值与*预测*值之间的差距。 

The loss function can quantify the difference between the actual value and the predicted value of the target.

通常我们会选择非负数作为损失，且数值越小表示损失越小，完美预测时的损失为0。`loss` 函数是表示神经网络性能的“恶劣程度”的指标，即当前的神经网络对监督数据在多大程度上不拟合，在多大程度上不一致。Loss 值越小模型越好。

`loss` 是关于未知参数的函数，不同的参数将算出不一样的 `loss`。

损失函数在参数空间中形成 `error surface`. error surface 是参数空间中的一个曲面，它描述了损失函数在不同参数取值下的变化情况。误差曲面的形状可以影响模型的训练过程，通常越平滑的误差曲面意味着模型训练的效率和稳定性更高，而越崎岖的误差曲面意味着模型训练过程更加困难。

<img width="458" alt="111" src="https://github.com/pillarliang/blog/assets/28833267/9f3bc1b5-d45f-4d08-9167-f67a7237b28a">

优化算法的目标是在误差曲面上寻找全局最优点或局部最优点，以最小化误差函数。然而，误差曲面通常是高维的、非凸的，并且存在多个局部最优点。因此，如何高效地在误差曲面上搜索最优解是深度学习中的重要问题。

# 常用的 `loss` 函数有：

## 回归

**绝对平均误差：`Mean absolute error` | MAE**

$$
Loss = {1\over N}\sum_{n}|y_n-\hat y_n|
$$

```python
torch.nn.L1Loss()
```

**均方误差：`Mean squared error`｜MSE**

衡量数据项相对于平均数的偏离程度。

- n 代表样本（数据的维数）

$$
Loss = {1\over N}\sum_{n}(y_n-\hat y_n)^2
$$

**标准差：`Standard Deviation`**

$$
SD = \sqrt{\frac{1}{n} \sum_{i=1}^{n} (x_i - \mu)^2}
$$

## 分类

**交叉熵误差：`Cross entropy error`**
<img width="606" alt="222" src="https://github.com/pillarliang/blog/assets/28833267/abcd15bf-2db9-4968-a362-2e371cde32ea">

它用于衡量模型的预测概率分布与真实标签的概率分布之间的差异。交叉熵误差越低，说明模型的预测越接近真实的标签。

以下是交叉熵误差的定义：

### **二分类问题:**

$H(y, p) = -\sum_{i=1}^{N} [y_i \log(p_i) + (1-y_i) \log(1-p_i)]$

其中：

- $y_i$ 是第 i 个样本的真实标签（0或1）。
- $p_i$ 是模型预测第 i 个样本的正类的概率。

### **多分类问题:**

$H(y, p) = -\sum_{i=1}^{N} \sum_{c=1}^{C} y_{ic} \log(p_{ic})$

其中：

- $y_{ic}$ 是第 i 个样本是否属于类别 c 的指示函数（如果属于则为1，否则为0）。
- $p_{ic}$ 是模型预测第 i 个样本属于类别 c 的概率。
